Improve surfacing of PythonEvaluator error messages.

anwar6953 opened the pull request:
https://issues.cask.co/browse/CDAP-4685

StringUtils surfaces the Python error messages in a better manner than the Log#error call, which also trims the actual python error.

anwar6953 committed:
Improve surfacing of PythonEvaluator error messages.

anwar6953 commented:
Previous Log message by ETLMapper:

```
2016-02-03 23:51:20,055 - ERROR [main:c.c.c.e.b.m.ETLMapReduce$ETLMapper@432] - Exception thrown in BatchDriver Mapper: {}
java.lang.IllegalArgumentException: Could not transform input: null
    at co.cask.hydrator.transforms.PythonEvaluator.transform(PythonEvaluator.java:177) ~[na:na]
    at co.cask.hydrator.transforms.PythonEvaluator.transform(PythonEvaluator.java:54) ~[na:na]
    at co.cask.cdap.etl.common.LoggedTransform$2.call(LoggedTransform.java:57) ~[cdap-etl-core-3.3.1-SNAPSHOT.jar:na]
    at co.cask.cdap.etl.common.LoggedTransform$2.call(LoggedTransform.java:54) ~[cdap-etl-core-3.3.1-SNAPSHOT.jar:na]
    at co.cask.cdap.etl.log.LogContext.run(LogContext.java:54) ~[cdap-etl-core-3.3.1-SNAPSHOT.jar:na]
    at co.cask.cdap.etl.common.LoggedTransform.transform(LoggedTransform.java:54) ~[cdap-etl-core-3.3.1-SNAPSHOT.jar:na]
    at co.cask.cdap.etl.common.TrackedTransform.transform(TrackedTransform.java:43) ~[cdap-etl-core-3.3.1-SNAPSHOT.jar:na]
    at co.cask.cdap.etl.common.TransformExecutor.executeTransformation(TransformExecutor.java:85) ~[cdap-etl-core-3.3.1-SNAPSHOT.jar:na]
    at co.cask.cdap.etl.common.TransformExecutor.executeTransformation(TransformExecutor.java:90) ~[cdap-etl-core-3.3.1-SNAPSHOT.jar:na]
    at co.cask.cdap.etl.common.TransformExecutor.runOneIteration(TransformExecutor.java:49) ~[cdap-etl-core-3.3.1-SNAPSHOT.jar:na]
    at co.cask.cdap.etl.batch.mapreduce.ETLMapReduce$ETLMapper.map(ETLMapReduce.java:406) ~[1454543468335-0/:na]
    at org.apache.hadoop.mapreduce.Mapper.run(Mapper.java:145) [hadoop-mapreduce-client-core-2.4.0.2.1.15.0-946.jar:na]
    at co.cask.cdap.internal.app.runtime.batch.MapperWrapper.run(MapperWrapper.java:102) [co.cask.cdap.cdap-app-fabric-3.3.1-SNAPSHOT.jar:na]
    at org.apache.hadoop.mapred.MapTask.runNewMapper(MapTask.java:764) [hadoop-mapreduce-client-core-2.4.0.2.1.15.0-946.jar:na]
    at org.apache.hadoop.mapred.MapTask.run(MapTask.java:340) [hadoop-mapreduce-client-core-2.4.0.2.1.15.0-946.jar:na]
    at org.apache.hadoop.mapred.YarnChild$2.run(YarnChild.java:168) [hadoop-mapreduce-client-app-2.4.0.2.1.15.0-946.jar:na]
    at java.security.AccessController.doPrivileged(Native Method) [na:1.7.0_75]
    at javax.security.auth.Subject.doAs(Subject.java:415) [na:1.7.0_75]
    at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1594) [hadoop-common-2.4.0.2.1.15.0-946.jar:na]
    at org.apache.hadoop.mapred.YarnChild.main(YarnChild.java:163) [hadoop-mapreduce-client-app-2.4.0.2.1.15.0-946.jar:na]
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.7.0_75]
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57) ~[na:1.7.0_75]
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.7.0_75]
    at java.lang.reflect.Method.invoke(Method.java:606) ~[na:1.7.0_75]
    at co.cask.cdap.internal.app.runtime.batch.distributed.MapReduceContainerLauncher.launch(MapReduceContainerLauncher.java:91) [launcher.jar:na]
    at org.apache.hadoop.mapred.YarnChild.main(Unknown Source) [hadoop-mapreduce-client-app-2.4.0.2.1.15.0-946.jar:na]
Caused by: org.python.core.PyException: null
    at org.python.core.Py.NameError(Py.java:246) ~[na:na]
    at org.python.core.PyFrame.getglobal(PyFrame.java:265) ~[na:na]
    at org.python.pycode._pyx0.transform$1(&lt;script&gt;:3) ~[na:na]
    at org.python.pycode._pyx0.call_function(&lt;script&gt;) ~[na:na]
    at org.python.core.PyTableCode.call(PyTableCode.java:165) ~[na:na]
    at org.python.core.PyBaseCode.call(PyBaseCode.java:166) ~[na:na]
    at org.python.core.PyFunction.__call__(PyFunction.java:338) ~[na:na]
    at org.python.pycode._pyx0.f$0(&lt;script&gt;:5) ~[na:na]
    at org.python.pycode._pyx0.call_function(&lt;script&gt;) ~[na:na]
    at org.python.core.PyTableCode.call(PyTableCode.java:165) ~[na:na]
    at org.python.core.PyCode.call(PyCode.java:18) ~[na:na]
    at org.python.core.Py.runCode(Py.java:1261) ~[na:na]
    at co.cask.hydrator.transforms.PythonEvaluator.transform(PythonEvaluator.java:174) ~[na:na]
    ... 25 common frames omitted
```

New error message by ETLMapper:

```
2016-02-04 01:40:51,123 - ERROR [main:c.c.c.e.b.m.ETLMapReduce$ETLMapper@432] - Exception thrown in BatchDriver Mapper: {}
java.lang.IllegalArgumentException: Could not transform input.
Traceback (most recent call last):
  File &quot;&lt;script&gt;&quot;, line 5, in &lt;module&gt;
  File &quot;&lt;script&gt;&quot;, line 2, in transform
TypeError: unsupported operand type(s) for +: 'long' and 'str'

    at org.python.core.Py.TypeError(Py.java:221)
    at org.python.core.PyObject._basic_add(PyObject.java:2017)
    at org.python.core.PyObject._add(PyObject.java:1995)
    at org.python.pycode._pyx0.transform$1(&lt;script&gt;:3)
    at org.python.pycode._pyx0.call_function(&lt;script&gt;)
    at org.python.core.PyTableCode.call(PyTableCode.java:165)
    at org.python.core.PyBaseCode.call(PyBaseCode.java:166)
    at org.python.core.PyFunction.__call__(PyFunction.java:338)
    at org.python.pycode._pyx0.f$0(&lt;script&gt;:5)
    at org.python.pycode._pyx0.call_function(&lt;script&gt;)
    at org.python.core.PyTableCode.call(PyTableCode.java:165)
    at org.python.core.PyCode.call(PyCode.java:18)
    at org.python.core.Py.runCode(Py.java:1261)
    at co.cask.hydrator.transforms.PythonEvaluator.transform(PythonEvaluator.java:171)
    at co.cask.hydrator.transforms.PythonEvaluator.transform(PythonEvaluator.java:55)
    at co.cask.cdap.etl.common.LoggedTransform$2.call(LoggedTransform.java:57)
    at co.cask.cdap.etl.common.LoggedTransform$2.call(LoggedTransform.java:54)
    at co.cask.cdap.etl.log.LogContext.run(LogContext.java:54)
    at co.cask.cdap.etl.common.LoggedTransform.transform(LoggedTransform.java:54)
    at co.cask.cdap.etl.common.TrackedTransform.transform(TrackedTransform.java:43)
    at co.cask.cdap.etl.common.TransformExecutor.executeTransformation(TransformExecutor.java:85)
    at co.cask.cdap.etl.common.TransformExecutor.executeTransformation(TransformExecutor.java:90)
    at co.cask.cdap.etl.common.TransformExecutor.runOneIteration(TransformExecutor.java:49)
    at co.cask.cdap.etl.batch.mapreduce.ETLMapReduce$ETLMapper.map(ETLMapReduce.java:406)
    at org.apache.hadoop.mapreduce.Mapper.run(Mapper.java:145)
    at co.cask.cdap.internal.app.runtime.batch.MapperWrapper.run(MapperWrapper.java:102)
    at org.apache.hadoop.mapred.MapTask.runNewMapper(MapTask.java:764)
    at org.apache.hadoop.mapred.MapTask.run(MapTask.java:340)
    at org.apache.hadoop.mapred.YarnChild$2.run(YarnChild.java:168)
    at java.security.AccessController.doPrivileged(Native Method)
    at javax.security.auth.Subject.doAs(Subject.java:415)
    at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1594)
    at org.apache.hadoop.mapred.YarnChild.main(YarnChild.java:163)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at co.cask.cdap.internal.app.runtime.batch.distributed.MapReduceContainerLauncher.launch(MapReduceContainerLauncher.java:91)
    at org.apache.hadoop.mapred.YarnChild.main(Unknown Source)

    at co.cask.hydrator.transforms.PythonEvaluator.transform(PythonEvaluator.java:174) ~[na:na]
    at co.cask.hydrator.transforms.PythonEvaluator.transform(PythonEvaluator.java:55) ~[na:na]
    at co.cask.cdap.etl.common.LoggedTransform$2.call(LoggedTransform.java:57) ~[cdap-etl-core-3.3.1-SNAPSHOT.jar:na]
    at co.cask.cdap.etl.common.LoggedTransform$2.call(LoggedTransform.java:54) ~[cdap-etl-core-3.3.1-SNAPSHOT.jar:na]
    at co.cask.cdap.etl.log.LogContext.run(LogContext.java:54) ~[cdap-etl-core-3.3.1-SNAPSHOT.jar:na]
    at co.cask.cdap.etl.common.LoggedTransform.transform(LoggedTransform.java:54) ~[cdap-etl-core-3.3.1-SNAPSHOT.jar:na]
    at co.cask.cdap.etl.common.TrackedTransform.transform(TrackedTransform.java:43) ~[cdap-etl-core-3.3.1-SNAPSHOT.jar:na]
    at co.cask.cdap.etl.common.TransformExecutor.executeTransformation(TransformExecutor.java:85) ~[cdap-etl-core-3.3.1-SNAPSHOT.jar:na]
    at co.cask.cdap.etl.common.TransformExecutor.executeTransformation(TransformExecutor.java:90) ~[cdap-etl-core-3.3.1-SNAPSHOT.jar:na]
    at co.cask.cdap.etl.common.TransformExecutor.runOneIteration(TransformExecutor.java:49) ~[cdap-etl-core-3.3.1-SNAPSHOT.jar:na]
    at co.cask.cdap.etl.batch.mapreduce.ETLMapReduce$ETLMapper.map(ETLMapReduce.java:406) ~[1454550039121-0/:na]
    at org.apache.hadoop.mapreduce.Mapper.run(Mapper.java:145) [hadoop-mapreduce-client-core-2.4.0.2.1.15.0-946.jar:na]
    at co.cask.cdap.internal.app.runtime.batch.MapperWrapper.run(MapperWrapper.java:102) [co.cask.cdap.cdap-app-fabric-3.3.1-SNAPSHOT.jar:na]
    at org.apache.hadoop.mapred.MapTask.runNewMapper(MapTask.java:764) [hadoop-mapreduce-client-core-2.4.0.2.1.15.0-946.jar:na]
    at org.apache.hadoop.mapred.MapTask.run(MapTask.java:340) [hadoop-mapreduce-client-core-2.4.0.2.1.15.0-946.jar:na]
    at org.apache.hadoop.mapred.YarnChild$2.run(YarnChild.java:168) [hadoop-mapreduce-client-app-2.4.0.2.1.15.0-946.jar:na]
    at java.security.AccessController.doPrivileged(Native Method) [na:1.7.0_75]
    at javax.security.auth.Subject.doAs(Subject.java:415) [na:1.7.0_75]
    at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1594) [hadoop-common-2.4.0.2.1.15.0-946.jar:na]
    at org.apache.hadoop.mapred.YarnChild.main(YarnChild.java:163) [hadoop-mapreduce-client-app-2.4.0.2.1.15.0-946.jar:na]
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.7.0_75]
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57) ~[na:1.7.0_75]
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.7.0_75]
    at java.lang.reflect.Method.invoke(Method.java:606) ~[na:1.7.0_75]
    at co.cask.cdap.internal.app.runtime.batch.distributed.MapReduceContainerLauncher.launch(MapReduceContainerLauncher.java:91) [launcher.jar:na]
    at org.apache.hadoop.mapred.YarnChild.main(Unknown Source) [hadoop-mapreduce-client-app-2.4.0.2.1.15.0-946.jar:na]
```

Note that the actual python errors in the two stack traces are different (one is for adding string to number and the other is for a variable that doesn't exist).

gsps1 commented on python-evaluator-transform/src/main/java/co/cask/hydrator/transforms/PythonEvaluator.java:
```diff
@@ -34,6 +34,7 @@
 import co.cask.hydrator.plugin.transform.ScriptContext;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Maps;
+import org.apache.hadoop.util.StringUtils;
```
did you want to use hadoop StringUtils class ?

anwar6953 replied:
That's the only StringUtils that has a `stringifyException` message.

gsps1 replied:
you can use `ExceptionUtils.getStackTrace(Throwable)`

anwar6953 replied:
Of `org.apache.commons.lang.exception`?

gsps1 replied:
yea

albertshau commented on python-evaluator-transform/src/main/java/co/cask/hydrator/transforms/PythonEvaluator.java:
```diff
@@ -174,7 +171,7 @@ public void transform(StructuredRecord input, final Emitter<StructuredRecord> em
 
     } catch (Exception e) {
-      throw new IllegalArgumentException("Could not transform input: " + e.getMessage(), e);
+      throw new IllegalArgumentException("Could not transform input.\n" + StringUtils.stringifyException(e));
```
Lets add a comment here to explain that PyException only exposes the actual cause if .printStackTrace() is called on it, which is why its not logged by ETLMapper

albertshau commented:
comment about adding a comment

anwar6953 commented:
Updated to use `org.apache.commons.lang3.exception.ExceptionUtils` instead of `org.apache.hadoop.util.StringUtils`.
Also added a comment about why we need to go about it this way.

albertshau commented:
:+1:
